lib/train/transports/ssh_connection.rb

```diff
@@ -101,6 +101,11 @@ def login_command
       LoginCommand.new("ssh", args)
     end
 
+    def connected?
+      return false if session.nil?
+      session.exec!("whoami").match?(/Please login as the user/) ? false : true
```

Comment:
Any time you have a conditional whose branches evaluate to straight booleans is an opportunity to remove the conditional entirely. For example:

```ruby
# original

return false if session.nil?
session.exec!("whoami").match?(/Please login as the user/) ? false : true

# abstract:

return false if x.nil?
x.y ? false : true

# avoid `nil?`:

return false if !x
x.y ? false : true

# combine the logic:

!x || x.y ? false : true

# flip the logic to put true first:

!(!x || x.y) ? true : false

# at this point: we don't need a conditional or the true/false:

!(!x || x.y)

# apply DeMorgan's Law to clean it up: !(a||b) -> !a&&!b

x && !x.y

# unabstract:

session && !session.exec!("whoami").match?(/Please login as the user/)
```

The final version actually reads like what it is trying to do: "If there is a session and running whoami doesn't match 'blah'".

I'd take it one step further and not use `match`:

```ruby
session && !session.exec!("whoami") =~ /Please login as the user/

# then push `!` down:

session && session.exec!("whoami") !~ /Please login as the user/
```

(but that is personal preference (but also bears out in benchmarks))